### PR TITLE
advancedtls: add fields for root and identity providers in API

### PR DESCRIPTION
**Pull request by ZhenLian**
This PR mainly aims to add options `IdentityProvider` under `IdentityCertificateOptions`(new struct, see below) and `RootProvider` under `RootCertificateOptions`  for users to use. Currently, each `*CertificateOptions` will contain three ways of provisioning corresponding certificates: static credential struct, reloading callback functions, and Provider. We also added a  pre-configuration check using reflection to enforce users to only specify one of them. 
If users specify Provider, we'll get the updated credentials through `Provider.KeyMaterial` call, and feed them to the per-connection callback-based reloading.
Besides the major changes, this PR also includes changes to:

1. use a new struct `IdentityCertificateOptions` to contain all the ways of provisioning identity certs(just like the existing `RootCertificateOptions` for provisioning root certs). Some field names and comments are also simplified and improved.
2. Add two tests `TestClientOptionsConfig` and  `TestServerOptionsConfig` for some pre-configuration checks. Some checks previously inside `TestClientServerHandshake` were checking against some preconditions, and they were all moved to the two new tests. 
3. instead of loading test credential files in each test, utilize the `certStore` defined in `advancedtls_integration_test.go`, which will avoid writing loading code in `TestClientServerHandshake`.

### CODE REVIEW

**Comment by ZhenLian** [comment]
@easwars This PR adds the Provider interfaces into advancedtls APIs. Could you please take a look? Thank you so much!

**Comment by ZhenLian** [comment]
@jiangtaoli2016 FYI.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:100]: Is this really required?
Why is clearly documenting which field takes priority when more than one field is set, not enough?

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:167]: Why is this embedding required? Just saving a few characters which referring to these options does not seem like a good enough reason to me, especially since it hurts readability of the code. This is especially more pronounced in the `RootCertificateOptions` case since it is not even at the top of the struct, so its not very obvious.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:236]: If we simply document the order of preference of the options, and then make sure that at least one of them is set, then we could use a simple switch statement to pick the highest priority option which is set while constructing the `tls.Config`.
The idea is here: https://play.golang.org/p/ru5-mzMbPiz

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:238]: This can block forever if the `Provider` does not provide any key material.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:239]: Given that there is no way for this context to be cancelled from outside, I don't see the need for this extra context. You could simply pass `context.Background()` to `KeyMaterial()`.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:256]: Please do not break up lines unless it is really required. Go does not impose any line length restrictions.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:260]: Same here.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:255]: Is there really a case where the provider can return no certs, and yet not return an error?

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:325]: Comments about the context and this call blocking forever apply here as well.

**Review by easwars — CHANGES_REQUESTED** [review]
Havent looked at the `advancedtls_test.go` file yet.

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go]: Yeah that's a fair suggestion. Removed the embedded structure.

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go]: Because of the concerns mentioned above, I think it might be slightly better to keep the check. But using switch does make the code clearer, and I've already updated the code to reflect the change.

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go:100]: Because in my opinion, there is no use case when users want to specify more than one field.  If users specify more than one field, it is most likely that they misunderstand some part of the API, and we probably want to point it out directly instead of proceeding with the implicit priorities.

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go:264]: Yeah I don't think it will be the case. But since the code below is getting the first element of the slice, I think it might be better to make sure the slice won't be empty here?

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go]: Can I know the possible ways to solve this? Maybe add a deadline to this request?
I think from the API's perspective, if users use a Provider without feeding in any contents, that might be a misuse of the API. Adding a deadline might mitigate the issue. I'd just like to know if there are any other alternatives.

File: security/advancedtls/advancedtls.go
Comment (easwars) [reply security/advancedtls/advancedtls.go:264]: Makes sense since you are getting the first element of the slice.

Not sure what you mean by ` I think it might be better to make sure the slice won't be empty here?`

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:267]: Or alteast you can make a single check instead of two, `len(km.Certs) != 1`, and error out.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:284]: Same comments here about combining assignment and comparison.

File: security/advancedtls/advancedtls.go
Comment (easwars) [inline security/advancedtls/advancedtls.go:304]: Get rid of this comment?

File: security/advancedtls/advancedtls_test.go
Comment (easwars) [inline security/advancedtls/advancedtls_test.go:59]: Maybe move all this initialization to a `newFakeProvider` method, which returns `(*fakeProvider, error)`

File: security/advancedtls/advancedtls_test.go
Comment (easwars) [inline security/advancedtls/advancedtls_test.go:72]: You can also just have a single fakeProvider. Something like:

```
type provType int
const (
  provTypeRoot provType = iota
  provTypeIdentity
)

type fakeProvider struct {
  pt provType 
  isClient bool
  wantMultiCert bool
  wantErr bool
}
```

File: security/advancedtls/advancedtls_test.go
Comment (easwars) [inline security/advancedtls/advancedtls_test.go:70]: s/fake/f/

File: security/advancedtls/advancedtls_test.go
Comment (easwars) [inline security/advancedtls/advancedtls_test.go:105]: Could you please remove the embedding here.

File: security/advancedtls/advancedtls_test.go
Comment (easwars) [inline security/advancedtls/advancedtls_test.go:158]: You could also consider breaking this test into two, one for the success cases and one for the failure cases. That will simplify the actual test code inside `t.Run`.

File: security/advancedtls/advancedtls_test.go
Comment (easwars) [inline security/advancedtls/advancedtls_test.go:185]: Similar comments as the other test.

File: security/advancedtls/advancedtls_test.go
Comment (easwars) [inline security/advancedtls/advancedtls_test.go:168]: Please identify the function which failed and its inputs in error messages.
```
t.Fatalf("ClientOptions{%v}.config() = %v, wantErr %v", clientOptions, err, test.wantErr)
```
Same comment applies to other error messages also.

File: security/advancedtls/advancedtls.go
Comment (easwars) [reply security/advancedtls/advancedtls.go]: I think it might be ok to not specify a deadline here. But this should be made obvious to the user. Because eventually, their `grpc.Dial()` will exceed its deadline, and they will call `Close` on their `Provider` which will unblock this call.

If you want to support adding a deadline here, that needs to be plumbed through the API. Maybe something for you to consider.

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go:101]: SG. Let's use the reflection to count non-nil fields and do the enforcement check for now. We can revisit this in the future if we have some real use cases that users have to specify more than one field.

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go:197]: `RequireClientCert` controls whether the client sends certificates to the server, and `RootOptions` controls whether the server verifies the client's certificates. These two are a bit different here, so I think we'd better keep both of them.

File: security/advancedtls/advancedtls_test.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls_test.go]: That's a great idea, but having an initialization makes it difficult to create temporary object and feed it directly to the test structs below. We can refactor the test code but that makes the test structs contains a bit more fields, since we need to include all the fields in fakeProvider to the test struct. 
As an alternative, I am considering having an `init()` function that will load all the test data into a global variable. Then this `newFakeProvider` can just return `*fakeProvider`  without error. This might go into a separate PR. What do you think about this idea?

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go]: SG. I think it's a requirement in our case to have initial bootstrapping credentials before we start connection, so no need to add  a deadline here. I will document in the comments saying explicitly that Provider needs to have some contents.

**Review by easwars — APPROVED** [review]
Looks good in general. But you might also want to get an approval from someone with more security expertise. Thanks.

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:87]: If multiple options are set (is it allowed?), how do we select between them?

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:86]: nit. s/one of the options/one option

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:115]: nit. s/one of the options/one option

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:121]: nit. s/identity certs in/an identity cert for

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:125]: nit. s/identity certs in/an identity cert for

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:116]: `IdentityCertificate` seems a little weird to me. How about `LeafCertificate` or `CertificateChain` depending on whether it contains intermediate certificates or not?

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:92]: Just curious, is there a reason for not caching the root certificates, rather than calling the callback for each connection?

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [reply security/advancedtls/advancedtls.go:116]: Or just using `Certificate`? Since it's sufficient to distinguish it from `RootCertificate`.

File: security/advancedtls/advancedtls.go
Comment (yihuazhang) [inline security/advancedtls/advancedtls.go:212]: It seems this logic is repeated in both client and server options. Can we de-duplicate it?

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go]: Reloading per connection is more flexible, and users can do whatever they want on top of that. Also, current implemented caching approach is for PEM files. if users don't want to reload from PEM files, they would need to implement their Provider implementations, but if their use cases are simple enough they can just choose callback approach. 
It's not difficult to expose it in Go, and can provide more choices, so I expose it here.

File: security/advancedtls/advancedtls.go
Comment (ZhenLian) [reply security/advancedtls/advancedtls.go:116]: I first saw the name "identity certificate" couple of times in some gRPC team's proposals, and I soon started to use it,  because I think it conveys the usage of this certificate(which is to present identity to the peers), just like "root certificate"(which represents root of trust). Now it is used in many places in C core...
From this perspective, a word like `Certificate` doesn't serve this purpose, because technically speaking, a root certificate is also a certificate. Can we keep the word if you don't mind? :)

Comment (yihuazhang) [reply security/advancedtls/advancedtls.go:116]: Ah Ok. If `identity certificate` is an agreed upon term within gRPC, I am fine with it.

**Review by yihuazhang — APPROVED** [review]

**Pull request merged**
